File: api/examples/python/render_pdf.py
import json
import os
import sys
import uuid
import urllib.request


BASE_URL = os.getenv("NETIFY_BASE_URL", "http://localhost:3002").rstrip("/")
API_KEY = os.getenv("NETIFY_API_KEY")
JOB_ID = os.getenv("JOB_ID", str(uuid.uuid4()))
PDF_URL = os.getenv("PDF_URL")
USER_ID = os.getenv("USER_ID", "demo-user")
PROMPT = os.getenv("NETIFY_PROMPT", "Create a clear narrated training video.")


def request(url, method="GET", body=None, stream=False):
    headers = {"Authorization": f"Bearer {API_KEY}"}
    if body is not None:
        headers["Content-Type"] = "application/json"

    req = urllib.request.Request(
        url,
        data=json.dumps(body).encode("utf-8") if body is not None else None,
        headers=headers,
        method=method,
    )
    return urllib.request.urlopen(req, timeout=None if stream else 60)


if not API_KEY:
    print("Set NETIFY_API_KEY to your Netify API key.", file=sys.stderr)
    sys.exit(1)

if not PDF_URL:
    print("Set PDF_URL to a public or signed PDF URL.", file=sys.stderr)
    sys.exit(1)


start = request(
    f"{BASE_URL}/api/render",
    method="POST",
    body={
        "jobId": JOB_ID,
        "pdfUrl": PDF_URL,
        "userId": USER_ID,
        "prompt": PROMPT,
    },
)

print("Job accepted:", start.read().decode("utf-8"))
print(f"Streaming progress for {JOB_ID}")

events = request(f"{BASE_URL}/api/events/{JOB_ID}", stream=True)

for raw in events:
    line = raw.decode("utf-8").strip()
    if not line.startswith("data: "):
        continue

    event = json.loads(line[6:])
    print(f"{event['pct']}% {event['message']}")

    if event["pct"] == 100:
        print("Done. Retrieve the final MP4 from your configured video storage.")
        break

    if event["pct"] < 0:
        raise RuntimeError(event["message"])

File: api/examples/node/render-pdf.mjs
import { randomUUID } from 'node:crypto'

const baseUrl = process.env.NETIFY_BASE_URL || 'http://localhost:3002'
const apiKey = process.env.NETIFY_API_KEY
const jobId = process.env.JOB_ID || randomUUID()
const pdfUrl = process.env.PDF_URL
const userId = process.env.USER_ID || 'demo-user'
const prompt = process.env.NETIFY_PROMPT || 'Create a clear narrated training video.'

if (!apiKey) {
  console.error('Set NETIFY_API_KEY to your Netify API key.')
  process.exit(1)
}

if (!pdfUrl) {
  console.error('Set PDF_URL to a public or signed PDF URL.')
  process.exit(1)
}

const authHeaders = { Authorization: `Bearer ${apiKey}` }

const start = await fetch(`${baseUrl}/api/render`, {
  method: 'POST',
  headers: {
    ...authHeaders,
    'Content-Type': 'application/json',
  },
  body: JSON.stringify({ jobId, pdfUrl, userId, prompt }),
})

if (!start.ok) {
  throw new Error(`Render request failed: HTTP ${start.status} ${await start.text()}`)
}

console.log('Job accepted:', await start.json())
console.log(`Streaming progress for ${jobId}`)

const events = await fetch(`${baseUrl}/api/events/${encodeURIComponent(jobId)}`, {
  headers: authHeaders,
})

if (!events.ok || !events.body) {
  throw new Error(`Event stream failed: HTTP ${events.status} ${await events.text()}`)
}

const reader = events.body.pipeThrough(new TextDecoderStream()).getReader()
let buffer = ''

while (true) {
  const { value, done } = await reader.read()
  if (done) break
  buffer += value

  const chunks = buffer.split('\n\n')
  buffer = chunks.pop() || ''

  for (const chunk of chunks) {
    const line = chunk.split('\n').find((part) => part.startsWith('data: '))
    if (!line) continue

    const event = JSON.parse(line.slice(6))
    console.log(`${event.pct}% ${event.message}`)

    if (event.pct === 100) {
      console.log('Done. Retrieve the final MP4 from your configured video storage.')
      process.exit(0)
    }

    if (event.pct < 0) {
      throw new Error(event.message)
    }
  }
}
